### ISSUE TITLE
Licensing

### ISSUE THREAD

**Issue by dannycolin** [OpenScienceMOOC/site#17]
Since we are going to remove the old bootstrap 3 code, I was wondering if you'll be ok if I license the new CSS code under the AGPLv3 license instead of MIT license. The AGPLv3 license is kind of the CC-BY-SA but for software code. That mean, any fork needs to be share-like (with the same license).

**Comment by Protohedgehog** [comment]
Would have to check with @Kevin-Mattheus-Moerman on that one :)

**Comment by Kevin-Mattheus-Moerman** [comment]
I'm not opposed to changing the license in principle. However, there are some points to consider. 

1. The link to the original code and developers
Websites I've built (e.g. kevinmoerman.org and gibboncode.org) started out as a clone from the [project-pages jekyll template](http://projectpages.github.io/project-pages/. ), which carries the MIT license. However I've implemented lots of changes as you can see, and now more and more changes are being implemented during development here. So, one could argue we've departed so much from that initial clone that the website currently is more "inspired by" rather than "developed from" that initial clone. Perhaps we can do a formal comparison between the original code and the current state. My current gut feeling is that we could decide on our own license independent of the jekyll template I used to inform this one. 
2. Could you explain your rationale for switching to the AGPLv3 license? What are your arguments against the MIT license? I think for open source software the "open source" nature is most important to me. So as long as the source is available and provided under a suitable (open source initiative approved) license we are in good shape. The choice of license however does influence a) interactions/integration with other packages, and b) interactions with the community. 
Copy left (GPL) licenses have sometimes hindered integration e.g. in the case of the use of libraries, as people fear their associated "Midas touch" might propagate into/"poison" their work. Therefore I know several open source enthusiasts who stay away from GPL for this reason. However I've also seen people get angry that companies clone permissive licensed (e.g. MIT/BSD) stuff, alter the code and subsequently create a closed source proprietary version they make money out of. They then feel like the company is making money out of their hard work. However there is a significant number of open source projects that benefits hugely from company involvement and for these especially the permissive licenses are better. 
I think there are fewer people that object to permissive licences then there are people that object to copy left licenses. In that respect I think we should really think about the implications of the switch. 
In our case I'm not sure if we could make a strong case for switching to a copy left license. I think we need not fear companies creating proprietary versions of our code. On the other hand I think it is not too difficult to image that members in our community may want to use/integrate this code in their own sites, running perhaps their own scripts (e.g. java content online). If we force them to switch their codes to AGPL because of this then they might look around at other projects first and avoid ours. 
I personally use GPL if I do not need or expect company involvement or commercial applications, and if the code is pretty much a stand-alone thing so I do not expect others to try to integrate it into their own work. However if I think people might want to integrate the code in their work, then I feel I can avoid a potential conflicts with licenses, and give them the flexibility of letting them choose their own license, by using a permissive license such as the MIT license. 
So in short, I feel permissive licenses are more suitable in this case.

**Comment by dannycolin** [comment]
Hey @Kevin-Mattheus-Moerman  thanks for your answer. 

1. > I've built (e.g. kevinmoerman.org and gibboncode.org) started out as a clone from the project-pages jekyll template, which carries the MIT license.

Since we've talked about removing the old Bootstrap 3, I created a clean branch where no MIT code from third-parties is present. We should be OK to relicense that branch.

2. > Could you explain your rationale for switching to the AGPLv3 license? 
What are your arguments against the MIT license?

My rationales for switching to AGPLv3 is exactly that I would like my code to 
stay in the "Commons". In other words, to make sure that no one is profiting
from the hard labour of the community. I mean I don't mind if someone use my
code for a commercial project (and the GPL doesn't limit anyone to do so) but
I would really like to be sure that any modification would benefit the community
as a whole. I think it's more in the spirit of Open Science than the MIT is.

> If we force them to switch their codes to AGPL because of this then they might 
look around at other projects first and avoid ours.

AGPL (or GPL) doesn't force you to license your code under the same license but only the
modification you do to the (A)GPL software. 

One good example could be the Linux Kernel. Android use it and they release the other
components of their Operating System under a more permissive license. An other 
example would be Wordpress. There's a lot of plugin under different licenses even 
proprietary ones and people are allowed to sell their software. Finally, one platform 
that a lot of people in the project seems to want to use for the MOOC is OpenEdX. That
platform use the AGPLv3 license and it doesn't seem to stop EdX from making money with
it. :)

**Comment by Kevin-Mattheus-Moerman** [comment]
Awesome work on the website btw! Perhaps I need to pull some of these changes into my other sites :smile: 

I'm aware that AGPLv3 code does not preclude commercialization efforts. However it does make it difficult and the GPL licenses do not in principle allow the codes to be incorporated into proprietary systems (as is explained [here](https://www.gnu.org/licenses/gpl-faq.en.html#GPLInProprietarySystem)).

Both Linux and Wordpress use GPLv2 I believe. I might need a law degree to figure out the detailed differences between v2 and v3. It does seem GPLV3 more explicitly says "The GNU General Public License does not permit incorporating your program into proprietary programs.". Also both Linux and Wordpress (and edX) do not make money from the codes, instead companies have been created around offering services/support products around the projects.  

You mention: 

> AGPL (or GPL) doesn't force you to license your code under the same license but only the modification you do to the (A)GPL software.

Correct, in some cases the change is not required.... but in some cases it is. This is the whole point of the "copy left" style license, and why it is often described as "viral", and why I used the term "Midas touch". 
Again, as is explained [here](https://www.gnu.org/licenses/gpl-faq.en.html#GPLInProprietarySystem):

> A system incorporating a GPL-covered program is an extended version of that program. The GPL says that any extended version of the program must be released under the GPL if it is released at all.

But exactly how to "tip toe" around the license is vague, and difficult to understand, which is why it may act as a deterrent for some: 

> However, in many cases you can distribute the GPL-covered software alongside your proprietary system. To do this validly, you must make sure that the free and nonfree programs communicate at arms length, that they are not combined in a way that would make them effectively a single program.
.....
The difference between this and “incorporating” the GPL-covered software is partly a matter of substance and partly form.


I'm largely aligned with :point_right:  [this post](http://nipy.sourceforge.net/nipy/stable/faq/johns_bsd_pitch.html)
The authors says: 

> ... the benefits of collaborating with the private sector are real, whereas the fear that some private company will “steal” your product and sell it in a proprietary application leaving you with nothing is not.

I would agree with that. Although personally I am much less interested in/concerned about involvement from the private sector (for this project), I want to leave this option open for others. So I am more concerned about being flexible enough to towards the entire open source community than I am concerned about a company "stealing" our stuff. 

This is why I would lean to permissive licenses for work where we want to be flexible to ensure we receive lots of community involvement. 

However, if the general consensus is a copy-left license is desirable then I rest my case and would be happy to support the switch to  AGPLv3. 

@Protohedgehog what are your thoughts on this.

**Comment by dannycolin** [comment]
First of all thanks for your time (especially digging in the somewhat not really user-friendly GPL FAQ). I must admit that I too feel like I need a law degree to understand all that stuff XD. 

So back to the main subject : 

> However, in many cases you can distribute the GPL-covered software alongside 
your proprietary system. To do this validly, you must make sure that the free 
and nonfree programs communicate at arms length, that they are not combined in 
a way that would make them effectively a single program.

I think there could be a compromise. We could only license the CSS and Javascript 
code under the AGPLv3 and everything else under the MIT.

The CSS and Javascript code will be replacing the functionnalities of Bootstrap but in addition the code will be tested to make sure it is accessible friendly. We could even separate that code into an other repository and only add the compiled version to the website. That way, It will make sure they're separated works.

That should be enough to satisfy the "arms lenght" because even if someone don't use the AGPLv3 code the website should still work (html doesn't depend on CSS or JS).

In addition to that, even if the CSS/JS code was linked, people could license their work under a license compatible with the GPL (e.g MIT License). You can refer to the section 
[Linking with GPL][linking] in the FAQ.

[linking]: https://www.gnu.org/licenses/gpl-faq.en.html#LinkingWithGPL

There's also a stackoverflow question on that (with links to the GPL FAQ) that explain it way better than I can. https://softwareengineering.stackexchange.com/questions/107883/agpl-what-you-can-do-and-what-you-cant

In a nutshell, It demonstrated the difference between "combined" works where you only include an unmodified librairy like I propose in my compromise and "derivated works".

> Although personally I am much less interested in/concerned about involvement from the private sector (for this project), I want to leave this option open for others. So I am more concerned about being flexible enough to towards the entire open source community than I am concerned about a company "stealing" our stuff.

At the same time, being "flexible" penalize the free software community because it gives the possibly to privatize the code. I know it's highly unlikely that a private corporation will do that but I think it's a political stand against the privatization of code. And I feel that it's more align with the spirit of Open Science. The goal is to make sure that knowledge will remain accessible, free and open. Sorry for my leftist propaganda on that one :P.

/me running in circle yelling : "Code is speech!". 
(https://steinhardt.nyu.edu/scmsAdmin/uploads/003/681/cuan_1036.pdf for those interested in the political clivage in the FOSS community. It's written by the Anthropologist Gabriella Coleman. She's well known for her work on the FOSS community, and other Internet-related stuff).

**Comment by Protohedgehog** [comment]
Oof, @Kevin-Mattheus-Moerman and @dannycolin thanks for the discussion on this! I think what we ultimately want is to maximise the dissemination of this code, whether for commercial purposes or not. If there are specific fragments of code that need to be licensed differently, I'm happy with that, as long as it doesn't make things too difficult for external users.

Speaking of OpenEdX, look what I just came across.. https://howto.p2pu.org/

**Comment by Kevin-Mattheus-Moerman** [comment]
@dannycolin @Protohedgehog I think a single license would be best and clearest. If after considering my points you want to go with AGPLv3 that is fine. 
![image](https://media1.tenor.com/images/2f7b5a25fc3356bb97e1334ff3c2c27e/tenor.gif)
No hard feelings :tulip: :rabbit:  :unicorn:  :rainbow:

BTW @Protohedgehog yes `:unicorn: :rainbow:` does produce :unicorn: :rainbow: in markdown!

**Comment by dannycolin** [comment]
> @Protohedgehog I think what we ultimately want is to maximise the dissemination of this code, whether for commercial purposes or not

I'll count that as a vote for MIT. So we'll keep the source under the MIT for now. @Kevin-Mattheus-Moerman same no hard feelings ;). On the contrary, I really like that we can have that kind of discussion without starting a flamewar.

**Comment by Protohedgehog** [comment]
Awesome, thanks both! So, @Kevin-Mattheus-Moerman, the question is how do we exploit this new knowledge for maximum shenanigans.. :)

**Comment by Protohedgehog** [comment]
@Kevin-Mattheus-Moerman Noob question, I tried to get the rainbows and unicorns to appear here.. but no luck! https://opensciencemooc.github.io/site/about/ #halp!
